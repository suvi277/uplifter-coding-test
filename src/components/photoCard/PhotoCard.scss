@import '../../_shared/styles/variable';
@import '../../_shared/styles/mixin';

.card {
    margin: 20px 15px;
    cursor: pointer;
    transition: transform 1s ease;
    width: 100%;

    .card-body {
        display: flex;
        flex-direction: column;
        justify-content: space-between;

        h5 {
            font-size: 14px;
        }

        p {
            margin: 0;
            display: inline;
            font-size: 12px;

            &:first-child,
            img {
                display: none;
            }
        }
    }

    &:hover {
        @include transform(scale(1.1));
    }

    .block-with-text {
        overflow: hidden;
        position: relative;
        line-height: 1.2em;
        max-height: 3.6em;
        margin-right: -1em;
        padding-right: 1em;
        width: 100%;
    }

    .block-with-text:before {
        content: '...';
        position: absolute;
        right: 0;
        bottom: 0;
    }

    .block-with-text:after {
        content: '';
        position: absolute;
        right: 0;
        width: 1em;
        height: 1em;
        margin-top: 0.2em;
        background: white;
    }

    @media screen and (min-width: 576px) {
        width: calc(100% / 2 - 30px);
        width: auto;
    }

    @media screen and (min-width: 768px) {
        width: calc(100% / 2 - 30px);
    }

    @media screen and (min-width: 1024px) {
        width: calc(100% / 4 - 30px);
    }
}
