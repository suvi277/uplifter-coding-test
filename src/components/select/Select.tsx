import React from 'react';
export interface SelectProps {
    allOptions: string[];
    label: string;
}

const Select: React.SFC<SelectProps> = ({ allOptions, label }) => {
    return (
        <div className="input-group input-group-sm mt-3">
            <div className="input-group-prepend">
                <label className="input-group-text">{label}</label>
            </div>
            <select className="custom-select">
                {allOptions.map((tag, index) => (
                    <option key={index} value={tag}>
                        {tag}
                    </option>
                ))}
            </select>
        </div>
    );
};

export default Select;
