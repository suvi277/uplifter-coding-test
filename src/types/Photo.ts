export interface Photo {
    title: string;
    ownername: string;
    description: {
        _content: string;
    };
    url_q: string;
    owner: string;
    id: string;
    tags: string;
}

interface ServiceLoading {
    status: string;
}

interface ServiceLoaded<Photo> {
    status: string;
    photos?: Array<Photo>;
}

interface ServiceError {
    status: string;
    error?: Error;
}

export interface BaseAction {
    type: string;
    payload?: any;
}

export type PhotoCollection<T> = ServiceLoading | ServiceLoaded<T> | ServiceError;
