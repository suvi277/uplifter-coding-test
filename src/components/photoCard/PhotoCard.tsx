import React from 'react';
import { Photo } from '../../types/Photo';
import './PhotoCard.scss';
import Select from '../select/Select';

const PhotoCard: React.FC<Photo> = ({ title, ownername, description, url_q, owner, tags, id }) => {
    const alltags = tags ? tags.split(' ') : [];

    return (
        <div className="card shadow-lg">
            <img className="card-img-top" src={url_q} alt={title} />
            <div className="card-body">
                <h5 className="card-title">
                    <a
                        href={`https://www.flickr.com/photos/${owner}/${id}`}
                        rel="noopener noreferrer"
                        title={title}
                        target="_blank"
                    >
                        {title}
                    </a>{' '}
                    by{' '}
                    <a
                        href={`https://www.flickr.com/photos/${owner}`}
                        rel="noopener noreferrer"
                        title={ownername}
                        target="_blank"
                    >
                        {ownername}
                    </a>
                </h5>
                {description && (
                    <div
                        className="block-with-text"
                        dangerouslySetInnerHTML={{
                            __html: description._content
                        }}
                    />
                )}
                {alltags.length > 1 && <Select allOptions={alltags} label="Tags" />}
            </div>
        </div>
    );
};

export default PhotoCard;
