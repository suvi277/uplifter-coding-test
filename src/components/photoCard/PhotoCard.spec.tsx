import React from 'react';
import { shallow, mount } from 'enzyme';
import { Photo } from '../../types/Photo';
import Select from '../select/Select';
import PhotoCard from './PhotoCard';

describe('<PhotoCard /> with default props', () => {
    const defaultProps = {
        title: 'sample title',
        ownername: 'sample owner name',
        description: {
            _content: 'some content'
        },
        url_q: 'url/path',
        owner: 'sampleowner',
        id: 'some id',
        tags: ''
    };
    const wrapper = shallow(<PhotoCard {...defaultProps} />);
    it('renders as expected', () => {
        expect(wrapper).toMatchSnapshot();
    });
    it('should exist card and shadow-lg classname', () => {
        expect(wrapper.find('.card.shadow-lg')).toHaveLength(1);
    });

    it('should insert the prop values to correct places ', () => {
        const props: Photo = {
                ...defaultProps,
                tags: 'tag1 tag2'
            },
            photoCardComp = mount(<PhotoCard {...props} />);

        expect(photoCardComp.find('.card-title a').first().render().text()).toEqual(props.title);
        expect(photoCardComp.find('.card-title a').last().render().text()).toEqual(props.ownername);
        expect(photoCardComp.find('.card-img-top').render().attr('src')).toEqual(props.url_q);
    });

    it('should call the Select component when tags are empty ', () => {
        const props: Photo = {
                ...defaultProps,
                tags: ''
            },
            photoCardComp = shallow(<PhotoCard {...props} />);

        expect(photoCardComp.exists(Select)).toBeFalsy();
    });

    describe('when tags are not empty ', () => {
        const props: Photo = {
            ...defaultProps,
            tags: 'tag1 tag2'
        };
        let photoCardComp: any, expectedtags: any;
        beforeEach(() => {
            photoCardComp = mount(<PhotoCard {...props} />);
            expectedtags = props.tags.split(' ');
        });
        it('should call the Select component', () => {
            expect(photoCardComp.exists(Select)).toBeTruthy();
        });
        it('should call the Select component with valid props', () => {
            expect(photoCardComp.find('Select').prop('allOptions')).toEqual(expectedtags);
            expect(photoCardComp.find('Select').prop('label')).toEqual('Tags');
        });
    });
});
