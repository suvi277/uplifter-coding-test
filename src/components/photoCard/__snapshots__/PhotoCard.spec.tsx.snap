// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PhotoCard /> with default props renders as expected 1`] = `ShallowWrapper {}`;

exports[`PhotoCard Component renders as expected 1`] = `ShallowWrapper {}`;

exports[`StreamCard Component renders as expected 1`] = `ShallowWrapper {}`;
